src/wallet/wallet.cpp

```diff
@@ -394,10 +394,17 @@ std::shared_ptr<CWallet> RestoreWallet(WalletContext& context, const fs::path& b
     }
 
     auto wallet_file = wallet_path / "wallet.dat";
-    fs::copy_file(backup_file, wallet_file, fs::copy_options::none);
+    std::shared_ptr<CWallet> wallet;
 
-    auto wallet = LoadWallet(context, wallet_name, load_on_start, options, status, error, warnings);
+    try {
+        fs::copy_file(backup_file, wallet_file, fs::copy_options::none);
 
+        wallet = LoadWallet(context, wallet_name, load_on_start, options, status, error, warnings);
+    } catch (const std::exception& e) {
```

Comment:
In commit "Bugfix: Wallet: Wrap RestoreWallet content in a try block to ensure exceptions become returned errors and incomplete wallet directory is removed" (b60a2ce138f4549e7233542e21cd96e0d8e0b5f8)

This might just be difference of opinion about defensive programming, but IMO it makes code harder to reason about and maintain if it is catching exceptions that are never expected to be thrown. I think it would be better to only catch fs::filesystem_exception not std::exception, and only catch exceptions if copy_file throws, not if LoadWallet throws. I also think an error text like "failed to copy wallet" is better than "unexpected exception." Possible suggestion:

```diff
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -394,17 +394,16 @@ std::shared_ptr<CWallet> RestoreWallet(WalletContext& context, const fs::path& b
     }
 
     auto wallet_file = wallet_path / "wallet.dat";
-    std::shared_ptr<CWallet> wallet;
-
+    status = DatabaseStatus::SUCCESS;
     try {
         fs::copy_file(backup_file, wallet_file, fs::copy_options::none);
-
-        wallet = LoadWallet(context, wallet_name, load_on_start, options, status, error, warnings);
-    } catch (const std::exception& e) {
-        wallet.reset();  // just in case
-        if (!error.empty()) error += Untranslated("\n");
-        error += strprintf(Untranslated("Unexpected exception: %s"), e.what());
+    } catch (const fs::filesystem_error& e) {
+        error += Untranslated(strprintf(("Failed to copy wallet: %s"), e.what()));
+        status = DatabaseStatus::FAILED_CREATE;
     }
+
+    auto wallet = status == DatabaseStatus::SUCCESS ? LoadWallet(context, wallet_name, load_on_start, options, status, error, warnings) : nullptr;
+
     if (!wallet) {
         fs::remove(wallet_file);
         fs::remove(wallet_path);
```

Reply:
The GUI cannot handle _any_ exception here, so either we catch everything, or we have to modify the GUI to check for other exceptions and "crash" cleanly... IMO it's strictly better to fail to load the wallet, than to kill the program (cleanly or not).

Reply:
I agree with @ryanofsky. `LoadWallet` shouldn't throw any exceptions, and having it in this try-catch makes it seem like it could. The places within `LoadWallet` which are expected to throw already have their own try-catch blocks so that it returns an error rather than throwing an exception.

Reply:
The whole point is to catch *unexpected* exceptions...

Reply:
I do agree with ryanofsky and achow101 here. I don't think that we want to add try/catch everywhere around the sources if the functions aren't meant to throw exceptions (if they throw, then we have a bug in our sources that has to be solved).

I think that a better design would be to create a general safe connection function for `QTimer::singleShot` calls.
Just made a quick draft of it in https://github.com/bitcoin-core/gui/pull/666.

Reply:
> The whole point is to catch _unexpected_ exceptions...

I think then it would be better to implement some overarching exception handling in the GUI rather than handling them individually like this.

Reply:
Anything higher level wouldn't know to just abort loading the wallet rather than kill the entire program.

Reply:
RPC seems to be able to handle unexpected exceptions fine?

Reply:
Changing how the GUI handles unexpected exceptions in general is outside the scope of this PR.